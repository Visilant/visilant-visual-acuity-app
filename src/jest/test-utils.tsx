import React from 'react'
import { Provider as ReduxProvider } from 'react-redux'
import { fireEvent, render, RenderOptions, screen } from '@testing-library/react-native'
import { WrapperProps } from './types'
import { ThemeProvider } from 'styled-components/native'
import { appTheme } from '@components/theme/theme'
import { LocalizationProvider } from '@localizations/Provider'
import { store } from '@store'

const Providers: React.FunctionComponent<WrapperProps> = ({ children }) => {
  return (
    <ReduxProvider store={store}>
      <ThemeProvider theme={appTheme}>
        <LocalizationProvider>{children}</LocalizationProvider>
      </ThemeProvider>
    </ReduxProvider>
  )
}

const customRender = <T,>(
  ui: React.ReactElement<unknown, string | React.JSXElementConstructor<T>>,
  options?: RenderOptions
) => {
  return render(ui, {
    wrapper: Providers,
    ...options
  })
}

export { fireEvent, customRender as render, screen }
