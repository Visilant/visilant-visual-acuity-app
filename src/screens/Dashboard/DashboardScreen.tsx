import React from 'react'

import {
  Body,
  Footer,
  Header,
  Icon,
  IconsEnum,
  SafeView,
  Screen,
  Text,
  TextStyles
} from '@components'

import { ResultCard } from './components/ResultCard'

import { FooterImage, FooterInstructionSection, IconFlexBox, StartTestButton } from './styles'
import { useStackNavigation } from '@screens/stack'
import { useAppDispatch, useAppSelector } from '@store'
import { hideDashboardInstructions, selectShowDashboardInstructions } from '@store/settings'
import { useUser } from '@hooks/useUser'
import { withErrorBoundary } from '@hocs/withErrorBoundary'
import {
  clearActiveResults,
  clearDeepLinkParams,
  setActiveExaminationById
} from '@store/examinations'
import { useSortedExaminations } from './utils/useSortedExaminations'

const DashboardScreen = () => {
  const examinations = useSortedExaminations()
  const showInstruction = useAppSelector(selectShowDashboardInstructions)
  const dispatch = useAppDispatch()
  const { userName } = useUser()
  const navigation = useStackNavigation()

  const handleStart = () => {
    navigation.navigate('Test', {})
    dispatch(clearActiveResults())
    dispatch(clearDeepLinkParams())
    dispatch(hideDashboardInstructions())
  }

  const handleOpenResult = (mobileId: string) => () => {
    dispatch(setActiveExaminationById({ mobileId }))
    navigation.navigate('Results')
  }

  return (
    <>
      <Screen noBg={false}>
        <SafeView>
          <Header>
            <Text raw text={userName} textStyle={TextStyles.Title} />
            <IconFlexBox>
              <Icon icon={IconsEnum.info} action={() => navigation.navigate('Info')} />
              <Icon icon={IconsEnum.iconSettings} action={() => navigation.navigate('Settings')} />
            </IconFlexBox>
          </Header>
          <Body smallGap>
            <Text namespace="screens" text={'dashboard.subTitle'} textStyle={TextStyles.Footer} />
            {examinations?.map(value => (
              <ResultCard
                key={value.mobileId}
                examination={value}
                onPress={handleOpenResult(value.mobileId)}
              />
            ))}
          </Body>
        </SafeView>
      </Screen>
      <Footer>
        {showInstruction && (
          <FooterInstructionSection>
            <Text
              namespace="screens"
              text={'dashboard.instruction'}
              textStyle={TextStyles.Footer}
            />
            <FooterImage source={require('@assets/dashboard_footer.png')} />
          </FooterInstructionSection>
        )}
        <StartTestButton
          namespace="screens"
          text="dashboard.button"
          capitalize={true}
          onPress={handleStart}
        />
      </Footer>
    </>
  )
}

export default withErrorBoundary(DashboardScreen)
