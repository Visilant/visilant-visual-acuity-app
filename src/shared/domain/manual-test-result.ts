export enum ManualTestResult {
  FingerCounting = 'FC',
  HandMovements = 'HM',
  LightPerception = 'LP',
  NoPerception = 'NP'
}
