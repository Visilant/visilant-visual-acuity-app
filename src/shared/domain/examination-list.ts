import { Examination } from './examination'

export class ExaminationList {
  constructor(private _values: Examination[]) {}
  get values(): Examination[] {
    return this._values
  }
}
