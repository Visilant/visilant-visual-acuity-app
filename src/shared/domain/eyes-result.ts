import { SnellenChartRatio } from './snellen-chart'
import { ManualTestResult } from '@shared/domain/manual-test-result'

export type EyeResult = SnellenChartRatio | ManualTestResult

export interface EyesResult {
  left?: EyeResult
  right?: EyeResult
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace EyesResult {
  export const create = (
    left: EyeResult | undefined,
    right: EyeResult | undefined
  ): EyesResult => ({ left, right })

  export const copy = (prev: EyesResult | undefined): EyesResult | undefined => {
    return prev ? { left: prev.left, right: prev.right } : undefined
  }

  export const withLeft = (
    prev: EyesResult | undefined,
    value: EyeResult | undefined
  ): EyesResult | undefined => {
    return create(value, prev?.right)
  }

  export const withRight = (
    prev: EyesResult | undefined,
    value: EyeResult | undefined
  ): EyesResult | undefined => {
    return create(prev?.left, value)
  }
}
