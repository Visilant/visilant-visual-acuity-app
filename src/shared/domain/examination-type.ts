import { EyesResult } from './eyes-result'

export enum ExaminationType {
  NormalRight = 'NormalRight',
  NormalLeft = 'NormalLeft',
  PinholeRight = 'PinholeRight',
  PinholeLeft = 'PinholeLeft'
}

export type ExaminationData = {
  patientName?: string
  comment?: string
  normal?: EyesResult
  pinhole?: EyesResult
}

export interface ExaminationResult extends ExaminationData {
  mobileCreatedAt?: Date
}

export interface ExaminationToSave extends ExaminationData {
  mobileId?: string
}
