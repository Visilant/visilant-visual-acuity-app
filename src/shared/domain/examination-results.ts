import { EyeResult, EyesResult } from '@shared/domain/eyes-result'
import { ExaminationType } from '@shared/domain/examination-type'

export class ExaminationResults {
  constructor(
    private _mobileId?: string,
    private _normal?: EyesResult,
    private _pinhole?: EyesResult
  ) {}

  get normal(): EyesResult | undefined {
    return this._normal
  }

  get pinhole(): EyesResult | undefined {
    return this._pinhole
  }
  get mobileId(): string | undefined {
    return this._mobileId
  }

  copy(): ExaminationResults {
    return new ExaminationResults(this._mobileId, this._normal, this._pinhole)
  }

  public setSnellen(type: ExaminationType, value: EyeResult | undefined) {
    if (type === ExaminationType.NormalLeft) {
      this._normal = EyesResult.withLeft(this._normal, value)
    } else if (type === ExaminationType.NormalRight) {
      this._normal = EyesResult.withRight(this._normal, value)
    } else if (type === ExaminationType.PinholeLeft) {
      this._pinhole = EyesResult.withLeft(this._pinhole, value)
    } else if (type === ExaminationType.PinholeRight) {
      this._pinhole = EyesResult.withRight(this._pinhole, value)
    }
  }
}
