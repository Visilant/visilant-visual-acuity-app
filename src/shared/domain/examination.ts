import { EyesResult } from '@shared/domain/eyes-result'
import { ExaminationResult } from '@shared/domain/examination-type'
import { UUID } from '@shared/uuid'

export class Examination {
  constructor(
    private _mobileId: string,
    private _patientName: string,
    private _comment: string,
    private _mobileCreatedAt: Date,
    private _normal: EyesResult | undefined,
    private _pinhole: EyesResult | undefined,
    private _isSynced: boolean = true
  ) {}

  static fromResults(results: ExaminationResult): Examination {
    const { patientName, comment, mobileCreatedAt, normal, pinhole } = results
    return new Examination(
      UUID.v4(),
      patientName || '',
      comment || '',
      mobileCreatedAt || new Date(),
      normal,
      pinhole,
      false
    )
  }

  get mobileId(): string {
    return this._mobileId
  }

  get patientName(): string {
    return this._patientName
  }

  get comment(): string {
    return this._comment
  }

  get normal(): EyesResult | undefined {
    return this._normal
  }

  get pinhole(): EyesResult | undefined {
    return this._pinhole
  }

  get mobileCreatedAt(): Date {
    return this._mobileCreatedAt
  }

  get isSynced(): boolean {
    return this._isSynced
  }
}
