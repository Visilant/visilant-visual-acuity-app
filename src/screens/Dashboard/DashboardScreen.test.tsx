import { render } from '@jest/test-utils'
import { ExaminationList } from '@shared/domain/examination-list'
import DashboardScreen from '@screens/Dashboard/DashboardScreen'
import * as store from '@store'

jest.mock('@components/Icon')
const mockResource = new ExaminationList([])
jest.mock('@hooks/useExamination', () => ({ useExamination: () => mockResource }))
jest.mock('@screens/stack', () => ({ useStackNavigation: () => ({ navigate: jest.fn() }) }))
jest.spyOn(store, 'useAppSelector').mockImplementation(() => true)
jest.mock('@hooks/useUser', () => ({ useUser: () => ({ userName: 'User Name' }) }))

describe('Dashboard screen', () => {
  test('matches the snapshot', () => {
    const onboarding = render(<DashboardScreen />)

    expect(onboarding).toMatchSnapshot()
  })
})
