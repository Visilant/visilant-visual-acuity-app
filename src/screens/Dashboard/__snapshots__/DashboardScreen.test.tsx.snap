// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard screen matches the snapshot 1`] = `
[
  <View
    noBg={false}
    style={
      {
        "backgroundColor": "#EDF1F4",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <View
      compact={false}
      insets={
        {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        }
      }
      style={
        {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingLeft": 30,
          "paddingRight": 30,
          "paddingTop": 60,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderBottomColor": "rgba(41,45,92,0.1)",
            "borderBottomStyle": "solid",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingBottom": 25,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#23255A",
                "fontFamily": "Caprasimo",
                "fontSize": 36,
                "lineHeight": 38,
                "marginBottom": 0,
                "marginLeft": 0,
                "marginRight": 0,
                "marginTop": 0,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              },
              "
  ",
              [Function],
              "
  font-family: Caprasimo;
  font-size: 36px;
  line-height: 38px;
",
            ]
          }
        >
          User Name
        </Text>
        <View
          style={
            {
              "display": "flex",
              "flexDirection": "row",
              "gap": 10,
            }
          }
        />
      </View>
      <RCTScrollView
        contentContainerStyle={
          {
            "flexDirection": "column",
            "flexGrow": 0,
            "gap": 15,
            "justifyContent": "flex-start",
            "paddingBottom": 35,
            "paddingTop": 30,
          }
        }
        style={
          {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          }
        }
      >
        <View>
          <Text
            style={
              [
                {
                  "color": "#23255A",
                  "fontFamily": "Libre-Franklin-Medium",
                  "fontSize": 26,
                  "lineHeight": 32,
                  "marginBottom": 0,
                  "marginLeft": 10,
                  "marginRight": 0,
                  "marginTop": 0,
                  "paddingBottom": 20,
                  "paddingLeft": 0,
                  "paddingRight": 0,
                  "paddingTop": 0,
                  "textAlign": "left",
                },
                "
  font-size: 26px;
  font-family: Libre-Franklin-Medium;
  line-height: 32px;
  text-align: left;
  padding: 0px 0px 20px;
  margin-left: 10px;
  ",
                [Function],
                "
",
              ]
            }
          >
            Your testing history
          </Text>
        </View>
      </RCTScrollView>
    </View>
  </View>,
  <View
    insets={
      {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      }
    }
    style={
      {
        "backdropFilter": "blur(5px)",
        "backgroundColor": "rgba(255,255,255,0.9)",
        "borderTop": "1px solid rgba(41,45,92,0.1)",
        "paddingBottom": 30,
        "paddingLeft": 30,
        "paddingRight": 30,
        "paddingTop": 20,
      }
    }
  >
    <View
      style={
        {
          "display": "flex",
          "flexdirection": "row",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#23255A",
              "fontFamily": "Libre-Franklin-Medium",
              "fontSize": 26,
              "lineHeight": 32,
              "marginBottom": 0,
              "marginLeft": 10,
              "marginRight": 0,
              "marginTop": 0,
              "paddingBottom": 20,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
              "textAlign": "left",
            },
            "
  font-size: 26px;
  font-family: Libre-Franklin-Medium;
  line-height: 32px;
  text-align: left;
  padding: 0px 0px 20px;
  margin-left: 10px;
  ",
            [Function],
            "
",
          ]
        }
      >
        To start, click
 this button
      </Text>
      <Image
        source={1}
        style={
          {
            "bottom": "0.5%",
            "height": 123,
            "maxWidth": "auto",
            "position": "absolute",
            "right": -30,
            "width": 187,
          }
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      size={1}
      style={
        [
          {
            "alignItems": "center",
            "alignSelf": "stretch",
            "backgroundColor": "#EF4823",
            "borderBottomLeftRadius": 31,
            "borderBottomRightRadius": 31,
            "borderTopLeftRadius": 31,
            "borderTopRightRadius": 31,
            "flexWrap": "nowrap",
            "height": 62,
            "justifyContent": "center",
            "overflow": "hidden",
            "paddingBottom": 12,
            "paddingLeft": 30,
            "paddingRight": 30,
            "paddingTop": 12,
          },
          {
            "flexGrow": 0,
            "marginTop": "auto",
          },
        ]
      }
      type="normal"
    >
      <View
        style={{}}
      />
      <Text
        props={
          {
            "size": 1,
            "type": "normal",
          }
        }
        style={
          [
            {
              "color": "#E6EEF4",
              "flexWrap": "nowrap",
              "fontFamily": "Libre-Franklin-Bold",
              "fontSize": 24,
              "lineHeight": 38,
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
              "position": "relative",
              "textAlign": "center",
              "top": -1,
            },
            "
  position: relative;
  top: ",
            [Function],
            ";
  color: ",
            [Function],
            ";
  font-family: Libre-Franklin-Bold;

  ",
            [Function],
            ";
  text-align: center;
  flex-wrap: nowrap;
",
          ]
        }
      >
        START A NEW TEST
      </Text>
    </View>
  </View>,
]
`;
